### ISSUE TITLE
False positive for possibly unbounded variable in context manager ExitStack()

### ISSUE THREAD

**Issue by tareqsiraj** [microsoft/pyright#3701]
**Describe the bug**
pyright seems to think a variable assigned in the `with` block for `contextlib.ExitStack()` is possibly unbounded where the same code in the `with` block for the type works.

**To Reproduce**
```
❯ pyright main.py
No configuration file found.
No pyproject.toml file found.
stubPath /home/xxxx/dev/scratch/pyright_test/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.260
/home/xxxx/dev/scratch/pyright_test/main.py
  /home/xxxx/dev/scratch/pyright_test/main.py:27:7 - error: "input2" is possibly unbound (reportUnboundVariable)
1 error, 0 warnings, 0 informations
Completed in 0.537sec

❯ pyright --version
pyright 1.1.260
```

**Expected behavior**
If `pyright` understands code block 1 in the example, it should understand code block 2 as well.

**Screenshots or Code**
```python
#!/usr/bin/python3

import contextlib
from typing import Iterator

class Foo:
    def bar(self) -> str:
        return "bar"

@contextlib.contextmanager
def make_foo() -> Iterator[Foo]:
        yield Foo()


# 1===================
with make_foo() as b:
    input1 = b.bar()

print(input1)  # works
# ====================

# 2===================
with contextlib.ExitStack() as stack:
    b = stack.enter_context(make_foo())
    input2 = b.bar()

print(input2)  # "input2" is possibly unbound (reportUnboundVariable)
# ====================
```

**VS Code extension or command-line**
Command line.

**Additional context**
n/a

**Comment by tareqsiraj** [comment]
I am sad about this but understand that `pyright` needs to take the safest assumption here. Thanks for the details.
